File: frontend/src/api/profiles.ts
import { apiFetch } from './client';

export interface ProfileListEntry {
  username: string;
  join_date: string;
  total_steps_all_time: number;
}

export interface ProfileListResponse {
  profiles: ProfileListEntry[];
}

export function getProfiles(): Promise<ProfileListResponse> {
  return apiFetch<ProfileListResponse>('/profiles');
}

export interface Profile {
  username: string;
  token: string;
  join_date: string;
}

export function createProfile(body: { username: string }): Promise<Profile> {
  return apiFetch<Profile>('/profiles', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

File: backend/app/services/profiles.py
"""Service layer for /api/profiles.

The users-index endpoint joins profiles with their all-time step total.
The `total_steps_all_time` calculation mirrors `get_user_summary` in
services/steps.py — per-CT-day MAX(total), summed across all days —
so the number matches what the Profile page already shows."""

from collections import defaultdict

from sqlalchemy import text
from sqlalchemy.engine import Connection

from backend.app.schemas.profiles import (
    ProfileListEntry,
    ProfileListResponse,
)
from backend.app.services.steps import _all_time_daily_max


def list_profiles(conn: Connection) -> ProfileListResponse:
    profile_rows = (
        conn.execute(
            text(
                "SELECT id, username, join_date FROM profiles "
                "ORDER BY username ASC"
            )
        )
        .mappings()
        .all()
    )
    daily_max = _all_time_daily_max(conn)
    totals: dict[int, int] = defaultdict(int)
    for (uid, _d), t in daily_max.items():
        totals[uid] += t
    entries = [
        ProfileListEntry(
            username=p["username"],
            join_date=p["join_date"],
            total_steps_all_time=totals.get(int(p["id"]), 0),
        )
        for p in profile_rows
    ]
    return ProfileListResponse(profiles=entries)
